server/src/main/java/io/crate/collections/accountable/AccountableList.java

```diff
@@ -0,0 +1,210 @@
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.collections.accountable;
+
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.RandomAccess;
+import java.util.function.LongConsumer;
+
+import org.apache.lucene.util.RamUsageEstimator;
+
+/**
+ * List implementation accounting for shallow memory usage.
+ * Based on {@link ArrayList} with some simplifications.
+ * Doesn't have concurrent modification check.
+ * Supports boundary checks since some features (lag/lead) rely on IndexOutOfBoundsException.
+ */
+public class AccountableList<T> extends AbstractList<T> {
+
+    private static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private final LongConsumer allocateBytes;
+    private Object[] elementData;
+    private int size;
+
+    public AccountableList(LongConsumer allocateBytes) {
+        this.allocateBytes = allocateBytes;
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA; // Empty now, will be accounted on the first growth.
+        allocateBytes.accept(4); // internal 'size' integer;
+    }
+
+    private static class SubList<T> extends AbstractList<T> implements RandomAccess {
+        private final AccountableList<T> root;
+        private final int offset;
+        private int size;
+
+        public SubList(AccountableList<T> root, int fromIndex, int toIndex) {
+            this.root = root;
+            root.allocateBytes.accept(NUM_BYTES_OBJECT_REF + 8); // Pointer, offset and size.
+            this.offset = fromIndex;
+            this.size = toIndex - fromIndex;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void sort(Comparator<? super T> c) {
+            // Arrays.sort might require up to n/ 2 temporary object references.
+            root.allocateBytes.accept(RamUsageEstimator.alignObjectSize((long) NUM_BYTES_OBJECT_REF * size / 2));
```

Comment:
The general contract is that we only account for memory which is retained by a structure, not for temporary objects.

E.g. if you sort multiple times, the retained memory will be stable, but this would end up accounting more each time.

Same further below in the `addAll`. The new array is only temporary the retained memory is already getting accounted within `grow`

Reply:
https://github.com/crate/crate/pull/16317